### PR TITLE
scaled-scene-buffer: implement buffer sharing mechanism

### PR DESCRIPTION
**Pull request by tokyo4j**
Suggested in #2352, combined with @Consolatis's work.

This PR adds `equal()` API in scaled-scene-buffer which can be used to share buffers for visually duplicated elements (e.g. scaled-font-buffer with the same text, font color and size). This PR itself isn't much beneficial as most `scaled-font-buffer`s are not duplicated, but this will greatly reduce the increased memory usage in #2352 as `scaled-rect-buffer`s used for backgrounds of menu and menu items will be visually the same.

### CODE REVIEW

File: src/common/scaled-font-buffer.c [hunk 1]
```diff
@@ -42,9 +40,33 @@ _destroy(struct scaled_scene_buffer *scaled_buffer)
...
+	struct scaled_font_buffer *b = scaled_buffer_b->data;
+
+	return str_equal(a->text, b->text)
+		&& a->max_width == b->max_width
+		&& str_equal(a->font.name, b->font.name)
+		&& a->font.size == b->font.size
+		&& a->font.slant == b->font.slant
+		&& a->font.weight == b->font.weight
+		&& !memcmp(a->color, b->color, sizeof(float[4]))
+		&& !memcmp(a->bg_color, b->bg_color, sizeof(float[4]))
```
Comment (Consolatis) [inline src/common/scaled-font-buffer.c:62]: `sizeof(a->color)` (or `b->color`) feels safer to me here in case we ever end up changing `float[4]` to something else.

File: src/common/scaled-scene-buffer.c [hunk 2]
```diff
@@ -41,29 +41,71 @@ _cache_entry_destroy(struct scaled_scene_buffer_cache_entry *cache_entry, bool d
...
+			continue;
 		}
+
+		/* Ensure self->width and self->height are set correctly */
+		self->width = scene_buffer->width;
+		self->height = scene_buffer->height;
+
+		wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
+		wlr_scene_buffer_set_dest_size(self->scene_buffer, self->width, self->height);
+		return;
```
Comment (Consolatis) [inline src/common/scaled-scene-buffer.c:108]: I still think we should create a local cache entry when reusing a buffer. Otherwise it will not be reused once the original owner is closed. Maybe we can just do a `struct wlr_buffer *wlr_buffer = NULL` on top, set it here, remove the `return` and then do
```c
static void
_update_buffer(struct scaled_scene_buffer *self, double scale) {
	[..]
	if (!wlr_buffer) {
		/* Create new buffer, will get destroyed along the backing wlr_buffer */
		struct lab_data_buffer *buffer = self->impl->create_buffer(self, scale);
		self->width = buffer ? buffer->logical_width : 0;
		self->height = buffer ? buffer->logical_height : 0;
		wlr_buffer = &buffer->base;
	}
	if (wlr_buffer) {
		/* Ensure the buffer doesn't get deleted behind our back */
		wlr_buffer_lock(wlr_buffer);
	}
	/* Create or reuse cache entry */
	[..]
	cache_entry->buffer = wlr_buffer;
	[..]
}
```

**Review by Consolatis — COMMENTED** [review]
> This PR itself isn't much beneficial as most `scaled-font-buffer`s are not duplicated

If you spawn multiple windows with the same title (e.g. foot) then it will actually reuse the title. But agree that that is not the main benefit for this PR, that would be scalable window icon / button / menu background lookups.

Another pair of eyes on this PR would be good as tokyo4j and I were both involved in writing it.

File: include/common/scaled-scene-buffer.h [hunk 3]
```diff
@@ -17,6 +17,10 @@ struct scaled_scene_buffer_impl {
 		(struct scaled_scene_buffer *scaled_buffer, double scale);
 	/* Might be NULL or used for cleaning up */
 	void (*destroy)(struct scaled_scene_buffer *scaled_buffer);
+	/* Returns true if the two buffers are visually the same */
+	bool (*equal)(struct scaled_scene_buffer *scaled_buffer_a,
+		struct scaled_scene_buffer *scaled_buffer_b);
+	struct wl_list cached_buffers;
```
Comment (johanmalm) [inline include/common/scaled-scene-buffer.h:23]: We shouldn't put this list in the `impl` struct (which is in effect the method dispatch vtable).

File: include/common/scaled-scene-buffer.h [hunk 4]
```diff
@@ -32,7 +36,8 @@ struct scaled_scene_buffer {
 	struct wl_listener destroy;
 	struct wl_listener output_enter;
 	struct wl_listener output_leave;
-	const struct scaled_scene_buffer_impl *impl;
+	struct scaled_scene_buffer_impl *impl;
+	struct wl_list link;
```
Comment (johanmalm) [inline include/common/scaled-scene-buffer.h:40]: We should say what linked list this is for. Looks like `other cached scaled-scene-buffers`.

File: src/common/scaled-scene-buffer.c [hunk 5]
```diff
@@ -49,21 +49,63 @@ _cache_entry_destroy(struct scaled_scene_buffer_cache_entry *cache_entry, bool d
...
+		 * If found in our local cache,
+		 * - self->width and self->height are already set
+		 * - wlr_scene_buffer_set_dest_size() has already been called
+		 */
+		return;
+	}
+
+	/* Search from other cached scaled-scene-buffers */
+	struct scaled_scene_buffer *scene_buffer;
+	wl_list_for_each(scene_buffer, &self->impl->cached_buffers, link) {
```
Comment (johanmalm) [inline src/common/scaled-scene-buffer.c:87]: I struggle to get my head around this. Does each scaled-scene-buffer contain a list of 'other scaled scene buffers' in the impl-struct?

File: src/common/scaled-scene-buffer.c [hunk 6]
```diff
@@ -177,6 +220,10 @@ scaled_scene_buffer_create(struct wlr_scene_tree *parent,
 	self->active_scale = 1;
 	self->drop_buffer = drop_buffer;
 	wl_list_init(&self->cache);
+	if (!impl->cached_buffers.next) {
```
Comment (johanmalm) [inline src/common/scaled-scene-buffer.c:223]: It is not clear to me what this does.

What does `!impl->cached_buffers.next` mean? Couldn't it mean that the list has never been initialized OR contains items. Feels like red flag to me.

**Comment by johanmalm** [comment]
I've had a read through, and have to admit that I find it hard to follow. You're two very clever people, so I feel sure there is soundness in the approach, but I would feel happier if I could snap into the thinking.

I've left a few comments to show where I get confused as a starting point for discussion.

File: src/common/scaled-scene-buffer.c (hunk 6 above)
Comment (Consolatis) [reply src/common/scaled-scene-buffer.c:223]: It basically just means that its calloc'd and was not initialized. An empty list points to itself for both `.prev` and `.next` and a list with items points to these items. Maybe we could add some `wl_list_maybe_init()` helper or something? Other ideas welcome, the issue here is that this is a shared list between all scaled-buffers of a given implementation (just the the scaled-font-buffer at the moment). See also my reply to the comment regarding the placement of the list itself above.

File: include/common/scaled-scene-buffer.h (hunk 3 above)
Comment (Consolatis) [reply include/common/scaled-scene-buffer.h:23]: An alternative approach I could think of is using a static var per impl and then defining something like `scaled_font_buffer_init()` method that has to be called from somewhere like `server.c` and then does something like
```c
static wl_list cache;
[..]
void
scaled_font_buffer_init() {
	wl_list_init(&cache);
}
```

We could then supply that `&cache` to `scaled_scene_buffer_create()` whenever the impl is creating a new scaled buffer. It would cost one more pointer per `struct scaled_scene_buffer` but would be an alternative. Although I personally think that this is slightly more code compared to storing the list in the `impl` struct and doing a lazy `wl_list_init()`.

Comment (johanmalm) [reply include/common/scaled-scene-buffer.h:23]: I think a static variable at the start of the `scaled-scene-buffer.c` would be by far the clearest approach for me.

We could just init it in `scaled_scene_buffer_create()` with a quick check to make sure we only do it the first time.

Comment (Consolatis) [reply include/common/scaled-scene-buffer.h:23]: The cache belongs to the implementation of a scaled-scene-buffer, e.g. only scaled-font-buffer.c for now but each new implementation might want its own cache.

 > We could just init it in scaled_scene_buffer_create() with a quick check to make sure we only do it the first time.

That was the idea with the `if (!list.next) { wl_list_init(list); }`

Comment (johanmalm) [reply include/common/scaled-scene-buffer.h:23]: Yeah, it's just a fine line between "clear code" that someone like me (as reviewer) quickly gets - and using "tricks" to enable optimization.

Comment (johanmalm) [reply include/common/scaled-scene-buffer.h:23]: Yep. Maybe just `init_wl_list()` :smile:

Comment (johanmalm) [reply include/common/scaled-scene-buffer.h:23]: Or... how about just add a comment to describe this.

```
/*
 * As this is a shared list, it should only be initialized the first time.
 * When just `calloc()d` and not yet initialized, `prev`/`next` that is used to avoid
 * initizing a list in use
 */
```

Comment (Consolatis) [reply include/common/scaled-scene-buffer.h:23]: Fine with me as well. If we ever see the need to do this in other places we can always add a helper later.

File: src/common/scaled-scene-buffer.c (hunk 2 above)
Comment (tokyo4j) [reply src/common/scaled-scene-buffer.c:108]: Yeah I agree. I didn't do it just because it will incur some overhead for iterating over caches of each buffers when creating the first buffer for a certain scale (`impl->equal()` always returns `true` but `find_cache_for_scale()` returns `NULL`). But the overhead will be negligible and I agree that we should not differentiate owner / borrower buffers.

File: src/common/scaled-scene-buffer.c (hunk 5 above)
Comment (Consolatis) [reply src/common/scaled-scene-buffer.c:87]: Each scaled-scene-buffer has an impl pointer which in turn has a single list for that implementation (e.g. scaled-font-buffer) where all scene_buffers it has ever created (and their scene nodes are still alive) are stored. This would not change if we use a static wl_list per implementation and provide a `impl->get_cache()` getter for it. Do you have a suggestion to state that more clearly in the comment?

File: include/common/scaled-scene-buffer.h (hunk 3 above)
Comment (tokyo4j) [reply include/common/scaled-scene-buffer.h:23]: So, is adding a comment describing the intent of `impl.cached_buffer` enough for now? I'm not sure I understood the discussion correctly...

Another approach for one-shot initialization of `wl_list` would be defining a macro like this:

```c
#define WL_LIST_INIT(head) ((struct wl_list){.prev = (head), .next = (head)})

static struct scaled_scene_buffer_impl impl = {
	.cached_buffers = WL_LIST_INIT(&impl.cached_buffers),
};
```

I personally prefer this approach as it moves the task of initialization from runtime to compile-time and  eliminates additional checks.

File: src/common/scaled-scene-buffer.c (hunk 5 above)
Comment (tokyo4j) [reply src/common/scaled-scene-buffer.c:87]: Exactly. I should have added more detailed comments about it.

File: src/common/scaled-font-buffer.c (hunk 1 above)
Comment (tokyo4j) [reply src/common/scaled-font-buffer.c]: Done

File: src/common/scaled-scene-buffer.c (hunk 2 above)
Comment (tokyo4j) [reply src/common/scaled-scene-buffer.c]: Done

File: include/common/scaled-scene-buffer.h (hunk 4 above)
Comment (tokyo4j) [reply include/common/scaled-scene-buffer.h]: Added a comment:
```
	/*
	 * This holds all the created scaled-*-buffers and is referenced for
	 * sharing backing buffers
	 */
```

File: src/common/scaled-font-buffer.c (hunk 1 above)
Comment (Consolatis) [reply src/common/scaled-font-buffer.c]: 2nd one should be `sizeof(a->bg_color)` (its `sizeof(a->color)` currently).

Comment (tokyo4j) [reply src/common/scaled-font-buffer.c]: Oops. Fixed.

File: include/common/scaled-scene-buffer.h (hunk 3 above)
Comment (tokyo4j) [reply include/common/scaled-scene-buffer.h]: OK. I moved to the static var approach and added the following line in `scaled-font-buffer.c`:
```c
static struct wl_list cached_buffers = WL_LIST_INIT(&cached_buffers);
```

File: include/common/scaled-scene-buffer.h [hunk 7]
```diff
@@ -32,7 +35,13 @@ struct scaled_scene_buffer {
 	struct wl_listener destroy;
 	struct wl_listener output_enter;
 	struct wl_listener output_leave;
-	const struct scaled_scene_buffer_impl *impl;
+	struct scaled_scene_buffer_impl *impl;
```
Comment (Consolatis) [inline include/common/scaled-scene-buffer.h:38]: Can be turned back into a `const`.

File: src/common/scaled-scene-buffer.c [hunk 8]
```diff
@@ -176,7 +232,9 @@ scaled_scene_buffer_create(struct wlr_scene_tree *parent,
 	self->impl = impl;
 	self->active_scale = 1;
 	self->drop_buffer = drop_buffer;
+	self->cached_buffers = cached_buffers;
 	wl_list_init(&self->cache);
+	wl_list_insert(self->cached_buffers, &self->link);
```
Comment (Consolatis) [inline src/common/scaled-scene-buffer.c:237]: I think we should either `assert(cached_buffers)` or check for `self->cached_buffers` in `_update_buffer()`. Personally I prefer the 2nd variant as that allows future implementations that don't want to get their things cached (e.g. they pass `NULL` as arg and don't provide `impl->equal()`).

Comment (Consolatis) [reply src/common/scaled-scene-buffer.c]: Regarding the `wl_list_remove(&self->link)` in `_handle_node_destroy()`: that can be solved by either checking for `self->cached_buffers` before calling or by `wl_list_init(&self->link)` in `_create()` on `!cached_buffers`.

Or with an alternative: a pattern often used in wayland protocol implementations within wlroots. E.g. if you have a `struct wl_list` which is either part of some list or not it can always be removed via `wl_list_remove()` if it has been initialized after removal.

Comment (tokyo4j) [reply src/common/scaled-scene-buffer.c]: I agree. Added checks for `self->cached_buffers` in `scaled_scene_buffer_create()` and `_update_buffer()`.

Comment (Consolatis) [reply src/common/scaled-scene-buffer.c]: `wl_list_remove(&self->link);` in `_handle_node_destroy` will crash. See above

Comment (tokyo4j) [reply src/common/scaled-scene-buffer.c]: Sorry, I missed your second comment. I copied your suggested code.

Comment (Consolatis) [reply src/common/scaled-scene-buffer.c]: I thought so thus the 2nd reply :)

**Comment by Consolatis** [comment]
Added some buffer tracking and tested: no issues detected. The title buffers are correctly reused when spawning a 2nd foot window. When then closing the first one and opening a third foot window it now reuses the buffer from the 2nd instance. Everything seems to work as expected. Code LGTM as well. Great work!

@johanmalm this is ready from my point of view but please say if there is anything missing or should be documented better.

**Comment by johanmalm** [comment]
LGTM. This is a great piece of work laying some very useful building blocks for the future.
